src/clippy-js-proxy.c

```diff
@@ -0,0 +1,675 @@
+    }
+
+  if(_js_run_printf (priv->webview, NULL, &result,
+                     "function clippify (obj) {"
+                     "  obj._proxy = {};"
+                     "  for (prop in obj) {"
+                     "    if (prop === '_proxy') continue;"
+                     "    obj._proxy[prop] = obj[prop];"
+                     "    let property = prop;"
+                     "    Object.defineProperty (obj, property, {"
+                     "      set (val) {"
+                     "        this._proxy[property] = val;"
+                     "        window.webkit.messageHandlers.clippy_notify.postMessage({ property: property, value: val });"
+                     "      },"
+                     "      get () {"
+                     "        return this._proxy[property];"
+                     "      }"
+                     "    });"
+                     "  }"
+                     "  return obj;"
+                     "}"
+                     "clippify (%s);",
+                     priv->object_name))
+    return;
+
+  clippy_js_proxy_derived_sync (CLIPPY_JS_PROXY_DERIVED (object),
+                                webkit_javascript_result_get_js_value (result));
+  webkit_javascript_result_unref (result);
+}
+
+static void
+clippy_js_proxy_derived_class_init (ClippyJsProxyDerivedClass *klass, gpointer data)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  gint n_props = 1;
+
+  object_class->finalize = clippy_js_proxy_derived_finalize;
+  object_class->get_property = clippy_js_proxy_derived_get_property;
+  object_class->set_property = clippy_js_proxy_derived_set_property;
+  object_class->constructed  = clippy_js_proxy_derived_constructed;
+
+  if (!data)
+    return;
+
+  /* Save properties for our class */
+  klass->properties = data;
+
+  /* Count how many properties we have */
+  while (klass->properties[n_props])
+    n_props++;
+
+  klass->n_props = n_props;
+
+  /* Finally, Install properties */
+  g_object_class_install_properties (object_class, n_props, klass->properties);
+}
+
+static GParamSpec *
+param_spec_new_from_jsc (const gchar *name, JSCValue *value)
+{
+  GParamSpec *pspec;
+
+  if (jsc_value_is_string (value))
+    {
+      g_autofree gchar *string = jsc_value_to_string (value);
+      pspec = g_param_spec_string (name, "", "",
+                                   string,
+                                   G_PARAM_READWRITE);
+    }
+  else if (jsc_value_is_boolean (value))
+    {
+      pspec = g_param_spec_boolean (name, "", "",
+                                    jsc_value_to_boolean (value),
+                                    G_PARAM_READWRITE);
+    }
+  else if (jsc_value_is_number (value))
+    {
+      pspec = g_param_spec_double (name, "", "",
+                                   -G_MAXDOUBLE, G_MAXDOUBLE,
+                                   jsc_value_to_double (value),
+                                   G_PARAM_READWRITE);
+    }
+  else
+    {
+      return NULL;
```

Comment:
If you initialize `pspec` to `NULL`, then you skip this whole `else` block.

Reply:
true, but that would be an extra assignment for most cases :)